README.md
[![License LGPL 3](https://img.shields.io/badge/license-LGPLv3-blue.svg)](http://www.gnu.org/licenses/lgpl-3.0.txt)
[![Build Status](https://travis-ci.org/deib-polimi/SPF2.svg?branch=master)](https://travis-ci.org/deib-polimi/SPF2)

[![Issue Stats](http://issuestats.com/github/deib-polimi/SPF2/badge/pr?style=flat)](http://issuestats.com/github/deib-polimi/SPF2)
[![Issue Stats](http://issuestats.com/github/deib-polimi/SPF2/badge/issue?style=flat)](http://issuestats.com/github/deib-polimi/SPF2)

# SPF2

![alt tag](https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/SPF2_header.png)

<br>

## Informations

A social smart space is a physical space where members, both humans and smart objects, can exploit
services provided by the others in a technology-mediated way.
Social Proximity Framework (SPF) is a software solution for the creation of this spaces,
where social identity and proximity is taken into account to offer personalized services and support
real-life interactions with digital ones.
SPF2 is a new major release of [THIS SOFTWARE](https://github.com/deib-polimi/SPF) that fully support Wi-Fi Direct protocol.

SPF2 requires Android 4.2 JellyBean MR1 (API 17) or higher. But I tested this new version only on 4.4.x KilKat, 5.x.x Lollipop and 6.0 Marshmallow. 
This choice is related to to the fact that in previous versions, Wi-Fi Direct was unstable and unreliable.

This repository contains SPFApp, SPFFramework and SPFWFDMiddleware as local projects.
Also, there are the two local modules called SPFLib and SPFShared. To be able to create apps based on SPF2, 
you should add these two latest modules as remote dependencies into your build.gradle.

SPFShared and SPFLib are available on JCenter and Maven Central. If you need examples of how to create SPF's applications
look these 3 repositories: 
- [SPFCouponingProvider demo](https://github.com/deib-polimi/SPF2CouponingProviderDemo)
- [SPFCouponingClient demo](https://github.com/deib-polimi/SPF2CouponingClientDemo)
- [SPFChat demo](https://github.com/deib-polimi/SPF2ChatDemo)

If you want to modify SPFLib and SPFShared, I suggest to run Sonatype Nexus OSS on your local machine 
and upload the compiled AAR's into this Maven server. Also, remember to update the version of these libraries, 
because Gradle caches dependencies automatically into ~/.gradle/caches/modules-2/files-2.1 directory.


## Releases

- *10/30/2015* - **SPF2** 2.0.1 - [Download](https://github.com/deib-polimi/SPF2/releases/tag/v.2.0.1)
- *10/27/2015* - **SPF2** 2.0.0 - [Download](https://github.com/deib-polimi/SPF2/releases/tag/v.2.0.0)
- *10/26/2015* - **SPF2** RC 1 - [Download](https://github.com/deib-polimi/SPF2/releases/tag/v.rc1)
- *10/22/2015* - **SPF2** Beta 4- [Download](https://github.com/deib-polimi/SPF2/releases/tag/v.beta4)
- *10/20/2015* - **SPF2** Beta 3 - [Download](https://github.com/deib-polimi/SPF2/releases/tag/v.beta3)
- *10/19/2015* - **SPF2** Beta 2 - [Download](https://github.com/deib-polimi/SPF2/releases/tag/v.beta2)
- *10/14/2015* - **SPF2** Beta 1 - [Download](https://github.com/deib-polimi/SPF2/releases/tag/v.beta1)
- *10/13/2015* - **SPF2** Alpha 4 - [Download](https://github.com/deib-polimi/SPF2/releases/tag/v.alpha4)
- *10/08/2015* - **SPF2** Alpha 3 - [Download](https://github.com/deib-polimi/SPF2/releases/tag/v.alpha3)
- *10/06/2015* - **SPF2** Alpha 2 - [Download](https://github.com/deib-polimi/SPF2/releases/tag/v.alpha2)
- *10/06/2015* - **SPF2** Alpha 1 - [Download](https://github.com/deib-polimi/SPF2/releases/tag/v.alpha1)


## Changelog

Changelog is available [HERE](https://raw.githubusercontent.com/deib-polimi/SPF2/master/CHANGELOG.md)


## Known issues to fix

- [ ] If discovery phase fails, create a temporized system to restart it automatically.
- [ ] When SPF enters in onConnectionInfoAvailable(), start a timer on the client. At the end of the countdown, if this device isn’t connected to the GO it should restart the discovery phase.
- [ ] Refactor MainActivity.java to reduce codelines.
- [ ] Updates the proximity switch in the GUI accordingly to the service status. For example, when Eternal Connect is working, you should update the switch. I suggest to use Broadcast receivers (not Local, but Remote declared into the AndroidManifest.xml). Use GroupInfosFragment as an example.
- [ ] Remove all ListViews and replace they with RecyclerView. Use GroupInfosFragment as an example.
- [ ] Fix Tab positioning for tablets/smartphones. You should find a way to display centered tabs that fill the entire screen width, but with text on a single line. Your solution should work also during screen rotation.
- [ ] Improve the standard mode into the Wi-Fi Direct Middleware.

If you want to do something, create a Pull Request! XD


## Images

<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/button_iconics.png" width="250">   
<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/drawer.png" width="250"> <br /> <br />
<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/notifications.png" width="250"> <br /> <br />
<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/tablet1.png" width="700"> <br /> <br />
<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/tablet2.png" width="700"> <br /> <br />
<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/tablet3.png" width="700"> <br /> <br />
<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/about_fragment.png" width="700"> <br /> <br />

<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/1.png" width="250"> <br /> <br />
<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/2.png" width="250"> <br /> <br />
<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/3.png" width="250"> <br /> <br />
<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/4.png" width="250"> <br /> <br />
<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/5.png" width="250"> <br /> <br />
<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/6.png" width="250"> <br /> <br />
<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/7.png" width="250"> <br /> <br /> 
<img align="center" src="https://raw.githubusercontent.com/deib-polimi/SPF2/master/repo_images/8.png" width="250">



## Official documentations

The official doc is composed by these files:
- (2015) SPF2 Doc [GitHub repository](https://github.com/deib-polimi/SPF2_Documentation) and [PDF](https://github.com/deib-polimi/SPF2_Documentation/releases/download/v1.0/SPF_documentation.pdf)
- (2014) Master Thesis of Jacopo Aliprandi and Dario Archetti available [HERE](http://hdl.handle.net/10589/106727)


## How to publish dependencies?

Change the information inside "private_bintray.properties" to be able to push dependencies on JFrog Bintray.

```
bintrayUser=your_username 
bintrayApiKey=your_apikey
bintrayGpgPassword=your_chosen_gpg_password
```

After that, follow this procedure:
- Remember to update the versions inside "build.gradle" of spflib and spfshared <br />
- Sync and build the project <br />
- In Android Studio Terminal type these commands into the SPF's root folder:
  
```
gradlew install
gradlew bintrayUpload
```
- Done!!! <br />


## How to create SPF's applications?

Add these dependencies into your local module's build.gradle:

```
dependencies {
compile 'it.polimi.spf:spflib:2.0.0.1@aar'
compile 'it.polimi.spf:spfshared:2.0.0.1@aar'
}
```

Examples:
- [SPFCouponingProvider demo](https://github.com/deib-polimi/SPF2CouponingProviderDemo) <br />
- [SPFCouponingClient demo](https://github.com/deib-polimi/SPF2CouponingClientDemo) <br />
- [SPFChat demo](https://github.com/deib-polimi/SPF2ChatDemo) <br />


## License
Copyright 2014 Jacopo Aliprandi, Dario Archetti<br>
Copyright 2015 Stefano Cappa

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG.md
**2.0.1**

1. Fixed gradle scripts to manage dependencies in the correct way.
2. Improve app icon



**2.0.0**

1. project moved from Eclipse to Android Studio.
2. SPFShared and SPFLib are remote dependencies into a maven repository
for external apps, but at the same time they are local projects
for SPF.
3. created a Gradle task to upload dependencies into a Sonatype Nexus
Server.
4. updated to Lollipop/Marshmallow Material Design using support libraries.
5. code cleanup using Butterknife, Otto and Project Lombok. Also, improved
architecture removing cyclic dependencies.
6. completely removed AllJoyn/AllSeen’s middleware. SPF2 has a pure
Wi-Fi Direct Middleware implementation.
7. notification into Notification Drawer updated with a custom layout
(using RemoteViews) adding an “X” button to stop the Proximity
Service.
8. fixed BUG ID=0: crashes related to profile image. Now I’m using
android-crop by soundcloud to pick and cut images. Also, I added
another library to create a circular image view.
21
9. added Travis CI to the GitHub repository.
10. fixed BUG ID=2: problem when I’m trying to save profile’s fields. I
replaced the listener with a TextWatcher.
11. fixed BUG ID=9: when SPFCouponingClient is closed, SPFFramework
receives and saves coupons ignoring my filter. To fix this problem, I
changed the filter to use the db, instead of the data saved in memory.
12. huge code refactoring and design improvements into Wi-Fi Direct Middleware.
13. fixed BUG ID=11: NullPointerException in GroupActor.
14. implemented a quick workaround to specify go intent and autonomous
mode from the GUI, using switches.
15. fixed BUG ID=12: “service is null” in middleware.
16. performance improvements for group creation’s procedure into the
middleware.
17. fixed Bounjour protocol logic into the middleware.
18. added Toast notification to explain the connection status.
19. fixed BUG ID=14: NullPointerException (but I don’t remember where).
20. using the unique identifier I created a quick logic to tell to clients if a
device is a GO or not. In fact, now GOs have ids that start with the
“AP” prefix.
21. fixed BUG ID=15: implemented a logic to connect only to “AP” devices.
22. fixed BUG ID=16: related to ID=15 but isolated into ServiceList class.
23. implemented Wi-Fi Direct’s Autonomous Mode for GOs.
24. implemented a new logic to manage in a better way the selection of
Socket’s ports.
25. log system updated and added code to catch all exceptions in the
middleware.
26. increased connection’s timeout on clients to reduce problems during
the connection procedure.
22
27. fixed BUG ID=17: in a specific part of the middleware, socket.close()
were called by the UI Thread causing a caught and ignored exception.
Fixed using Otto’s events to create asynchronous calls.
28. implemented a complete Event’s hierarchy used for Otto bus.
29. completely redesigned GroupActor and its subclasses. The superclass
is a Thread and all subclasses implement run() and other abstract
methods.
30. removed ServerSocketAcceptor and moved into GroupOwnerActor.
31. removed cyclic dependency between WifiDirectMiddleware and WFDBroadcastReceiver
using Otto’s events.
32. implemented Wi-Fi Direct Group’s management into Wi-Fi Direct
middleware (not really stable, but the problem is related to the Android’s
implementation of this protocol).
33. fixed BUG ID 18: Android 6.0 Marshmallow changed the permission
policy. For this reason, I implemented a workaround to guide the user
to enable a specific feature, required by external apps to interact with
the framework.
34. SPFApp updated with the new Google Design library, for example to
create Tabs.
35. fixed BUG ID=20: added a default photo profile.
36. fixed BUG ID=21: fixed small bug related to tabs.
37. implemented Eternal Connect in a singleton class. It’s an evolution of
the Eternal Discovery that I previously created for Pigeon Messenger
(https://github.com/deib-polimi/PigeonMessanger). Eternal Connect
is able to discovery and connect automatically to nearby devices.
38. added two new fragments: GO Infos and About. The first shows all
informations about services and connection status, the second one is
an about page.
39. completely rewritten the navigation drawer using MaterialDrawer and
Iconics to create a modern GUI. Also, I reused the same code to update
the Multi-pane layout for tablets.
40. added two Broadcast receivers to update the GUI with available services
and devices.
23
41. created a new app icon in Material Design.

sPFApp/src/main/java/it/polimi/spf/app/LoadersConfig.java
/*
 * Copyright 2015 Stefano Cappa
 *
 * This file is part of SPF.
 *
 * SPF is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * SPF is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for
 * more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with SPF.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package it.polimi.spf.app;

/**
 * Class that contains all ids and constants related to loaders.
 * Created by Stefano Cappa on 20/10/15.
 */
public class LoadersConfig {

    //Convention used: every class has id in the form XY.
    //Where X is the class number and Y is the loader's number inside the class X.
    //It's only a convention, but this prevent to use the same id for different loader creating many problems.

    /**
     * Constants used in {@link it.polimi.spf.app.fragments.profile.ProfileFragment}
     */
    public static final int LOAD_PROFILE_LOADER_ID = 0;
    public static final int SAVE_PROFILE_LOADER_ID = 1;

    /**
     * Constants used in {@link it.polimi.spf.app.fragments.personas.PersonasFragment}
     */
    public static final int CREATE_PERSONA_LOADER = 10;
    public static final int LOAD_PERSONAS_LOADER = 11;
    public static final int DELETE_PERSONA_LOADER = 12;
    public static final String EXTRA_PERSONA = "persona";

    /**
     * Constants used in {@link it.polimi.spf.app.fragments.contacts.PeopleFragment}
     */
    public static final int LOAD_CONTACTS_LOADER = 20;
    public static final int LOAD_REQUEST_LOADER = 21;

    /**
     * Constants used in {@link it.polimi.spf.app.fragments.contacts.CircleFragment}
     */
    public static final int LOAD_CIRCLE_LOADER = 30;
    public static final int ADD_CIRCLE_LOADER = 31;
    public static final int DELETE_CIRCLE_LOADER = 32;
    public static final String EXTRA_CIRCLE = "circle";

    /**
     * Constants used in {@link it.polimi.spf.app.fragments.NotificationFragment}
     */
    public final static int MESSAGE_LOADER_ID = 40;
    public final static int MESSAGE_DELETER_ID = 41;
    public static final String EXTRA_MESSAGE_ID = "messageId";

    /**
     * Constants used in {@link it.polimi.spf.app.fragments.appmanager.AppManagerFragment}
     */
    public static final int APP_LOADER = 50;

    /**
     * Constants used in {@link it.polimi.spf.app.fragments.ActivityFragment}
     */
    public static final int LOAD_LIST_LOADER_ID = 60;
}
